### PR TITLE
[camera] enableAudio parameter on availableCameras

### PR DESCRIPTION
**Pull request by omartinma**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

Microphone permissions are just requested if needed.

*List which issues are fixed by this PR. You must list at least one issue.*

https://github.com/flutter/flutter/issues/112277

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [ ] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_web/lib/src/camera_web.dart [hunk 1]
```diff
@@ -99,8 +101,8 @@ class CameraPlugin extends CameraPlatform {
       // Request video and audio permissions.
       final html.MediaStream cameraStream =
           await _cameraService.getMediaStreamForOptions(
-        const CameraOptions(
-          audio: AudioConstraints(enabled: true),
+        CameraOptions(
+          audio: AudioConstraints(enabled: enableAudio),
```
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera_web.dart:105]: Wouldn't this be much simpler if we just did:
```
  audio: AudioConstraints(enabled: false),
```
in the web implementation?

**Comment by ditman** [comment]
(Questions coming from https://github.com/flutter/flutter/issues/112277#issuecomment-1266880620)

> * Should be enableAudio exposed so the user can manipulate it?
>   * In that case, should be mandatory?
>   * default to false?

I think the simplest solution here is, in the web, to pass `false` as the audio constraint. No need to expose it as a bool through the whole plugin interface.

The flow goes:

> Hey, tell me what cameras you have!
< (Requests permissions to use cameras): There you go
> Hey, give me this camera, with/without audio
< (If needed, requests permissions to use microphone): There you go

I think this works best from a granular permissions standpoint; only request permissions really when needed. This might be considered a breaking change for some users on the web, so we should point it out in the CHANGELOG and README very clearly.

### However...

If we want to leave the user an option to request all permissions in a single prompt, then:

> * Should be enableAudio exposed so the user can manipulate it?

Yes

>   * In that case, should be mandatory?

No

>   * default to false?

It should default to `true`, so it the existing behavior is kept (at least on the web? how does this work in other platforms?)

TL;DR: I think that the goal if we expose the new `enableAudio` bool in the `enumerateCameras` method is to do it in such a way that people that don't change their code at all, keep seeing the best experience.

**Comment by omartinma** [comment]
@ditman  I agree with all the comments above, much simpler. So then the flow is in 2 steps:

- Give me all the cameras you have.
- Set up the specific camera I want, and in case I want the microphone I will request permission for it.

Then with this flow, my question is, should we simplify the implementation of some methods on the web like 'getMediaStreamForOptions'. Right now we are not using the parameters for CameraOptions at all, and we just simply expose it because platform methods like 'getUserMedia' are exposing these parameters. It seems a bit counterintuitive to me from developer's perspective. It will not affect the usage of the API but it will be cleaner to read the implementation. What do you think?

**Comment by omartinma** [comment]
@ditman PR updated and here. Video with the flow: 



https://user-images.githubusercontent.com/11978998/194008048-2d74456e-5a14-4355-a414-1e59f94f86a8.mov


Now something that I'm missing is:
> It should default to true, so it the existing behavior is kept (at least on the web? how does this work in other platforms?)

- Are all platforms having the same 2 steps flow to request permissions?
- Should be documented somewhere that the parameter enableAudio on CameraController is the one that will request the microphone permission?

**Comment by maciejbrzezinski** [comment]
@ditman any updates here?

**Comment by stuartmorgan-g** [comment]
We've just completed the migration of the plugin code to the flutter/packages repository, as described in https://flutter.dev/go/flutter-plugins-repo-migration, and this repository is now being archived. Unfortunately that means that all in-progress PRs here must be moved to flutter/packages.

Please see [our instructions](https://github.com/flutter/flutter/wiki/Migrating-Plugins-repository-PRs-to-Packages) for an explanation of how to move your PR, and if you have any issues moving your PR please don't hesitate to reach out in the #hackers-ecosystem channel in [Discord](https://github.com/flutter/flutter/wiki/Chat).

Our apologies that your PR was caught in this one-time transition. We're aware that it's disruptive in the short term, and appreciate your help in getting us to a better long-term state!
